### PR TITLE
Implementation of recommendation 3.10 - local variables declaration

### LINKED ISSUE #155: Recommendation 3.10: Variables are declared close to the line where they are first used
Local variables are declared close to the point where they are first used. THis will minimize their scope.
Local variable declarations are usually initialized or initialized immediately after the declaration. The member fields of the class should be declared collectively.

### PR DESCRIPTION
**Pull request by petertrr**
## Which rule and warnings did you add?

This pull request closes #155

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] Updated rules-config.json
* [x] Updated available-rules.md

## Fixme

A lot of limitations when working with AST only.

### CODE REVIEW

File: diktat-rules/src/main/resources/rules-config.json [hunk 1]
```diff
@@ -342,5 +342,10 @@
     "name": "ENUMS_SEPARATED",
     "enabled": true,
     "configuration": {}
+  },
+  {
```
Comment (akuleshov7) [inline diktat-rules/src/main/resources/rules-config.json:346]: will need to merge conflicts

Comment (akuleshov7) [reply diktat-rules/src/main/resources/rules-config.json:346]: also if you will add your rule to diktat-analysis.yml - please add comments of what does your rule do

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt [hunk 2]
```diff
@@ -78,6 +78,7 @@ enum class Warnings(private val canBeAutoCorrected: Boolean, private val warn: S
     WRONG_WHITESPACE(true, "incorrect usage of whitespaces for code separation"),
     TOO_MANY_CONSECUTIVE_SPACES(true, "too many consecutive spaces"),
     ENUMS_SEPARATED(true, "split enumeration error"),
+    LOCAL_VARIABLE_EARLY_DECLARATION(false, "local variables need to be declared close to line where they are first used"),
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt:81]: "should be" declared ... to "the" line

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/identifiers/LocalVariablesRule.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ * 2. It is allowed to declare variables in outer scope compared to usage scope. It could be useful to store state, e.g. between loop iterations.
+ *
+ * Current limitations due to usage of AST only:
+ * * Only properties without initialization or initialized with expressions based on constants are supported.
+ * * Properties initialized with constructor calls cannot be distinguished from method call and are no supported.
+ */
+class LocalVariablesRule : Rule("local-variables") {
+    companion object {
+        private var functionInitializers = listOf(
+                "emptyList", "emptySet", "emptyMap", "emptyArray", "emptySequence",
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/identifiers/LocalVariablesRule.kt:44]: linkedMapOf?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/identifiers/LocalVariablesRule.kt [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+    private fun KtCallExpression?.isWhitelistedMethod() =
+            this?.run {
+                (referenceExpression() as KtNameReferenceExpression).getReferencedName() in functionInitializers &&
+                        valueArguments.isEmpty()
+            } ?: false
+
+    private fun warnMessage(name: String, declared: Int, used: Int) = "$name is declared on line $declared and used for the first time on line $used"
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/identifiers/LocalVariablesRule.kt:159]: and "is" used, please also separate line number and name by some separators like "<>" or "[]"

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/PsiUtils.kt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+/**
+ * Checks if this [PsiElement] is an ancestor of [block].
+ * Nodes like `IF`, `TRY` are parents of `ELSE`, `CATCH`, but their scopes are not intersecting, and false is returned in this case.
+ */
+fun PsiElement.isContainingScope(block: KtBlockExpression): Boolean {
+    when (block.parent.node.elementType) {
+        ElementType.ELSE -> getParentOfType<KtIfExpression>(true)
+        ElementType.CATCH -> getParentOfType<KtTryExpression>(true)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/PsiUtils.kt:91]: our code style recommends to make else -> branch

File: diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/LocalVariablesWarnTest.kt [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+                    |        println()
+                    |    }
+                    |}
+                """.trimMargin()
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.LOCAL_VARIABLE_EARLY_DECLARATION)
+    fun `local variables used only in this scope - positive example with comments`() {
```
Comment (akuleshov7) [inline diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/LocalVariablesWarnTest.kt:67]: need to add test that has a property in a class and same variable name in function and lambda. Something like this:
```
class E {
val a = ""
fun foo() {
val a = ""
listOf().foreach { a -> println(a) }
```

**Review by akuleshov7 — APPROVED** [review]
generally looks good, small comments inside

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/PsiUtils.kt (hunk 5 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/PsiUtils.kt:91]: Yes, it has disappeared somehow, thx for noticing

**Pull request merged**